cirq/_compat_test.py

```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
+    # second import it might be the aliasing loader already
+    assert importlib.util.find_spec("cirq.testing._compat_test_data.fake_a")
+
+
+def _import_parent_use_constant_from_deprecated():
+    """to ensure that module initializations set attributes correctly"""
+    # sets up the DeprecationFinders
+    import cirq.testing._compat_test_data
+
+    # the parent should have fake_a set on it as an attribute - just like
+    # a regular module import (e.g. cirq.ops)
+    # should have a DUPE_CONSTANT as its imported from the dupe submodule
+    assert cirq.testing._compat_test_data.fake_a.DUPE_CONSTANT == False
+
+
+def _import_deprecated_sub_use_constant():
+    """to ensure that submodule initializations set attributes correctly"""
+    # sets up the DeprecationFinders
+    import cirq.testing._compat_test_data.fake_a.dupe
+
+    # should have a DUPE_CONSTANT as its defined on it, set to False
+    assert cirq.testing._compat_test_data.fake_a.dupe.DUPE_CONSTANT == False
+
+
+def _import_deprecated_same_name_in_earlier_subtree():
+    from cirq.testing._compat_test_data.fake_a.sub.subsub.dupe import DUPE_CONSTANT
+
+    assert DUPE_CONSTANT
+
+
+def _import_top_level_deprecated():
+    from cirq.testing._compat_test_data.fake_freezegun import api
+
+    assert api
+
+
+# this is where the deprecation error should show where the deprecated usage
+# has occured, which is this file
+_deprecation_origin = ['_compat_test.py:']
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_a_deprecation_msg = [
+    'fake_a was used but is deprecated',
+    'Use cirq.testing._compat_test_data.module_a instead',
+] + _deprecation_origin
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_b_deprecation_msg = [
+    'fake_b was used but is deprecated',
+    'Use cirq.testing._compat_test_data.module_a.module_b instead',
+] + _deprecation_origin
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_ops_deprecation_msg = [
+    'fake_ops was used but is deprecated',
+    'Use cirq.ops instead',
+] + _deprecation_origin
+
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_freezegun_deprecation_msg = [
+    'fake_freezegun was used but is deprecated',
+    'Use freezegun instead',
+] + _deprecation_origin
+
+
+def _trace_unhandled_exceptions(*args, **kwargs):
+    from multiprocessing import Queue
+
+    queue: Queue = kwargs['q']
+    func = kwargs['func']
+    del kwargs['q']
+    del kwargs['func']
+    try:
+        func(*args, **kwargs)
+        queue.put(None)
+    # coverage: ignore
+    except BaseException as ex:
+        # coverage: ignore
+        msg = str(ex)
+        queue.put((type(ex).__name__, msg, traceback.format_exc()))
+
+
+def subprocess_context(test_func):
+    """Ensures that sys.modules changes in subprocesses won't impact the parent process."""
+    import os
+
+    ctx = multiprocessing.get_context("spawn" if os.name == 'nt' else "fork")
```

Comment:
Did you test on Windows?

Nit: Double quotes vs. single quotes.

Reply:
Yes, without this it was failing on Windows (which we do have in our test matrix).